components/cam/src/physics/cam/shoc.F90

```diff
@@ -2053,148 +2055,245 @@ subroutine shoc_tke(&
+  real(rtype), intent(in) :: brunt(shcol,nlev)
+  ! PBL height [m]
+  real(rtype), intent(in) :: pblh(shcol)
+  ! Heights on the mid-point grid [m]
+  real(rtype), intent(in) :: zt_grid(shcol,nlev)
+  ! Monin-Okbukov length [m]
+  real(rtype), intent(in) :: obklen(shcol)
 
-      ! Compute the return to isotropic timescale
-      isotropy(i,k)=min(maxiso,tscale1/(1._rtype+lambda*buoy_sgs_save*tscale1**2))
+  ! intent-inout
+  ! eddy coefficient for heat [m2/s]
+  real(rtype), intent(inout) :: tkh(shcol,nlev)
+  ! eddy coefficient for momentum [m2/s]
+  real(rtype), intent(inout) :: tk(shcol,nlev)
+  ! turbulent kinetic energy [m2/s2]
+  real(rtype), intent(inout) :: tke(shcol,nlev)
 
-      ! Dimensionless Okukhov length considering only 
-      !  the lowest model grid layer height to scale
-      z_over_L = zt_grid(i,nlev)/obklen(i)
+  ! intent-out
+  ! Return to isotropic timescale [s]
+  real(rtype), intent(out) :: isotropy(shcol,nlev)
 
-      if (z_over_L .gt. zL_crit_val .and. (zt_grid(i,k) .lt. pblh(i)+pbl_trans)) then
-        ! If surface layer is moderately to very stable, based on near surface 
-        !  dimensionless Monin-Obukov use modified coefficients of 
-        !  tkh and tk that are primarily based on shear production 
-        !  and SHOC length scale, to promote mixing within the PBL
-        !  and to a height slighty above to ensure smooth transition.
-        tkh(i,k)=Ckh_s*(shoc_mix(i,k)**2)*sqrt(sterm_zt(i,k))
-        tk(i,k)=Ckm_s*(shoc_mix(i,k)**2)*sqrt(sterm_zt(i,k))
-      else
-        ! Default definition of eddy diffusivity for heat and momentum  
-      
-        tkh(i,k)=Ckh*isotropy(i,k)*tke(i,k)
-        tk(i,k)=Ckm*isotropy(i,k)*tke(i,k)
-      endif
+  !local variables
+  integer :: icol, klev 
+  real(rtype) :: z_over_L
 
-      tke(i,k) = max(mintke,tke(i,k))
+  !parameters
+  real(rtype), parameter :: lambda_low   = 0.001_rtype
+  real(rtype), parameter :: lambda_high  = 0.04_rtype
+  real(rtype), parameter :: lambda_slope = 0.65_rtype
+  real(rtype), parameter :: brunt_low    = 0.02_rtype
 
-    enddo ! end i loop
-  enddo ! end k loop
+  ! Critical value of dimensionless Monin-Obukhov length
+  real(rtype), parameter :: zL_crit_val = 100.0_rtype
+  ! Transition depth [m] above PBL top to allow 
+  !   stability diffusivities
+  real(rtype), parameter :: pbl_trans = 200.0_rtype
+
+  ! Turbulent coefficients
+  real(rtype), parameter :: Cs  = 0.15_rtype
+  real(rtype), parameter :: Ck  = 0.1_rtype
+  real(rtype), parameter :: Ckh = 0.1_rtype
+  real(rtype), parameter :: Ckm = 0.1_rtype
+  real(rtype), parameter :: Ckh_s = 1.0_rtype
+  real(rtype), parameter :: Ckm_s = 1.0_rtype
+  real(rtype), parameter :: Ce  = Ck**3/Cs**4
+
+  real(rtype), parameter :: Ce1 = Ce/0.7_rtype*0.19_rtype
+  real(rtype), parameter :: Ce2 = Ce/0.7_rtype*0.51_rtype
+
+  real(rtype), parameter :: Cee = Ce1 + Ce2
+
+  real(rtype) :: a_prod_bu, a_prod_sh, a_diss, buoy_sgs_save, tscale, lambda
+
+  do klev = 1, nlev
```

Comment:
@singhbalwinder The way I see it, this loop really contains computations for three separate processes that could be broken into their own subroutines.  

Lines 2236 to 2250 should remain in this appropriately named adv_sgs_tke routine.

Lines 2252 to 2269 concerns the computation of the return to isotropy term and could be it's own routine after this one.  

Lines 2271 to 2289 concerns the computation of the eddy diffusivities and should be it's own subroutine.

In addition, note that I feel that line 2289 is out of place (a lingering my fault issue).  I feel this threshold check should be included in line 2247.  However, this will change answers so probably best to implement these changes in a b4b manner and I can fix this after the fact in a separate PR.